src/preload.py

```diff
@@ -42,116 +45,371 @@
+        return self.partition_table.image
+
+    @property
+    def end(self):
+        # last sector (included)
+        return self.start + self.size - 1
+
+    @property
+    def start_bytes(self):
+        return self.start * SECTOR_SIZE
+
+    @property
+    def size_bytes(self):
+        return self.size * SECTOR_SIZE
+
+    @property
+    def end_bytes(self):
+        # last byte (included)
+        return self.start_bytes + self.size_bytes - 1
+
+    def is_included_in(self, other):
+        return (
+            other.start <= self.start <= other.end and
+            other.start <= self.end <= other.end
+        )
 
+    def is_extended(self):
+        return self.partition_table.label == "dos" and self.type == "f"
 
-@contextmanager
-def losetup_context_manager(partition_name, image=None):
-    device = losetup_partition(partition_name, image=image)
-    yield device
-    losetup("-d", device)
+    def is_last(self):
+        # returns True if this partition is the last on the disk
+        return self == self.partition_table.get_partitions_in_disk_order()[-1]
 
+    def get_sfdisk_line(self):
+        result = "{} : start={}, size={}, type={}".format(
+            self.node,
+            self.start,
+            self.size,
+            self.type
+        )
+        if self.uuid is not None:
+            result += ", uuid={}".format(self.uuid)
+        if self.name is not None:
+            result += ', name="{}"'.format(self.name)
+        if self.bootable:
+            result += ", bootable"
+        return result
+
+    def losetup_context_manager(self):
+        return losetup_context_manager(
+            self.image,
+            self.start_bytes,
+            self.size_bytes,
+        )
 
-def expand_file(path, additional_space):
-    with open(path, "a") as f:
-        size = f.tell()
-        f.truncate(size + additional_space)
-
-
-def expand_ext4(partition_name, image=None):
-    # Resize ext4 filesystem
-    error = False
-    with losetup_context_manager(partition_name, image=image) as loop_device:
-        log.info("Using {}".format(loop_device))
-        log.info("Resizing filesystem")
-        try:
-            status = e2fsck("-p", "-f", loop_device, _ok_code=[0, 1, 2])
-            status = status.exit_code
-            if status == 0:
-                log.info("e2fsck: File system OK")
-            else:
-                log.warning("e2fsck: File system errors corrected")
-        except ErrorReturnCode:
-            log.error("e2fsck: File system errors could not be corrected")
-            error = True
-        resize2fs("-f", loop_device)
-    if error:
-        exit(1)
+    def mount_context_manager(self):
+        return mount_context_manager(
+            self.image,
+            self.start_bytes,
+            self.size_bytes,
+        )
+
+    def str(self):
+        return "partition n°{} of {}".format(self.number, self.image)
+
+    def _resize_last_partition_of_disk_image(self, additional_bytes):
```

Comment:
The fact that `_resize_last_partition_of_disk_image` and `_resize_partition_on_disk_image` both have the same `(self, additional_bytes)` method-signature looks a little confusing. Not that it really matters, but would these methods make "more sense" on a `Disk` class, rather than a `Partition` class?

Reply:
This is because they were part of Partition.resize().
This method got too large and I had to split it.
I still find them too long but couldn't find a way to split them more without reducing readability.

I don't understand the second comment. We don't have a Disk class (but we have PartitionTable). We want to resize the partition so Partition.resize seems to be the good place. Note that a Partition knows the PartitionTable it belongs to.

Reply:
My apologies - I was commenting while tired and got confused ;-)
"Note that a Partition knows the PartitionTable it belongs to" is the bit I didn't realise.

Reply:
No problem, feedback is always appreciated.